"panic: reflect: call of reflect.Value.Elem on slice Value" in abi.go when calling bound contract method that returns 2 uint8 values

#### System information

Geth version: `1.5.0-unstable`
OS & Version: OS X 10.11.3 (15D21)

#### Expected behaviour

Return two numbers from this contract method from a contract file created with abigen

Contract Method:

```go
// TemperatureRange is a free data retrieval call binding the contract method 0x22aab53a.
//
// Solidity: function temperatureRange() constant returns(uint8, uint8)
func (_TemperatureMeasurementA *TemperatureMeasurementACaller) TemperatureRange(opts *bind.CallOpts) (uint8, uint8, error) {
	var (
		ret0 = new(uint8)
		ret1 = new(uint8)
	)
	out := []interface{}{
		ret0,
		ret1,
	}
	err := _TemperatureMeasurementA.contract.Call(opts, out, "temperatureRange")
	value := reflect.ValueOf(out)
	typ := value.Type()
	fmt.Println(typ)
	return *ret0, *ret1, err
}
```

I call the method through the following session:

```go
	session := &TemperatureMeasurementASession {
		Contract: tmaContract,
		CallOpts: bind.CallOpts{
			Pending: false,
		},
		TransactOpts: bind.TransactOpts{
			From:     auth.From,
			Signer:   auth.Signer,
			GasLimit: big.NewInt(3141592),
		},
	}
```

With this code:

```go
	minTemp, maxTemp, err := session.TemperatureRange()
```

#### Actual behaviour

Panics (See backtrace)

#### Steps to reproduce the behaviour

1. Create a solidity contract with a constant method that returns two uint8 numbers (see method above)

2. Create abi file test.abi with browser solidity

3. Create binary code test.bin with browser solidity

4. Generate a contract binding file with the command:
`abigen --abi test.abi --pkg main --type TestContract --out test.go --bin test.bin`

5. Deploy the contract using go (https://github.com/ethereum/go-ethereum/wiki/Native-DApps:-Go-bindings-to-Ethereum-contracts#deploying-contracts-to-ethereum)

6. Call the method using go

#### Backtrace

````
panic: reflect: call of reflect.Value.Elem on slice Value

goroutine 1 [running]:
panic(0x4489e20, 0xc8302da6a0)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
reflect.Value.Elem(0x43cbf00, 0xc8302da560, 0x97, 0x0, 0x0, 0x0)
	/usr/local/go/src/reflect/value.go:735 +0x228
github.com/ethereum/go-ethereum/accounts/abi.ABI.Unpack(0x0, 0x0, 0x0, 0xc8200f9860, 0x4, 0x4, 0x0, 0x0, 0x0, 0xc820296840, ...)
	/Users/tim/Documents/go_workspace/src/github.com/ethereum/go-ethereum/accounts/abi/abi.go:241 +0x18d
github.com/ethereum/go-ethereum/accounts/abi/bind.(*BoundContract).Call(0xc8200ee090, 0xc82029fef0, 0x43cbf00, 0xc8302da560, 0x45d0c40, 0x10, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/tim/Documents/go_workspace/src/github.com/ethereum/go-ethereum/accounts/abi/bind/base.go:108 +0x263
main.(*TemperatureMeasurementACaller).TemperatureRange(0xc8200b3a90, 0xc82029fef0, 0xc8302d6480, 0x0, 0x0)
	/Users/tim/Documents/go_workspace/src/eth_test/TemperatureMeasurementA.go:331 +0x1cd
main.(*TemperatureMeasurementASession).TemperatureRange(0xc82029fee8, 0xc8302d6480, 0x0, 0x0)
	/Users/tim/Documents/go_workspace/src/eth_test/TemperatureMeasurementA.go:342 +0x4a
main.main()
	/Users/tim/Documents/go_workspace/src/eth_test/main.go:128 +0xbf6
exit status 2
````

--- a/accounts/abi/abi.go
+++ b/accounts/abi/abi.go
@@ -238,8 +238,16 @@ func (abi ABI) Unpack(v interface{}, name string, output []byte) error {
 		return fmt.Errorf("abi: unmarshalling empty output")
 	}
 
-	value := reflect.ValueOf(v).Elem()
-	typ := value.Type()
+	// make sure the passed value is a pointer
+	valueOf := reflect.ValueOf(v)
+	if reflect.Ptr != valueOf.Kind() {
+		return fmt.Errorf("abi: Unpack(non-pointer %T)", v)
+	}
+
+	var (
+		value = valueOf.Elem()
+		typ   = value.Type()
+	)
 
 	if len(method.Outputs) > 1 {
 		switch value.Kind() {
@@ -268,6 +276,25 @@ func (abi ABI) Unpack(v interface{}, name string, output []byte) error {
 				return fmt.Errorf("abi: cannot marshal tuple in to slice %T (only []interface{} is supported)", v)
 			}
 
+			// if the slice already contains values, set those instead of the interface slice itself.
+			if value.Len() > 0 {
+				if len(method.Outputs) > value.Len() {
+					return fmt.Errorf("abi: cannot marshal in to slices of unequal size (require: %v, got: %v)", len(method.Outputs), value.Len())
+				}
+
+				for i := 0; i < len(method.Outputs); i++ {
+					marshalledValue, err := toGoType(i, method.Outputs[i], output)
+					if err != nil {
+						return err
+					}
+					reflectValue := reflect.ValueOf(marshalledValue)
+					if err := set(value.Index(i).Elem(), reflectValue, method.Outputs[i]); err != nil {
+						return err
+					}
+				}
+				return nil
+			}
+
 			// create a new slice and start appending the unmarshalled
 			// values to the new interface slice.
 			z := reflect.MakeSlice(typ, 0, len(method.Outputs))
@@ -293,34 +320,6 @@ func (abi ABI) Unpack(v interface{}, name string, output []byte) error {
 		}
 	}
 
-	return nil
-}
-
-// set attempts to assign src to dst by either setting, copying or otherwise.
-//
-// set is a bit more lenient when it comes to assignment and doesn't force an as
-// strict ruleset as bare `reflect` does.
-func set(dst, src reflect.Value, output Argument) error {
-	dstType := dst.Type()
-	srcType := src.Type()
-
-	switch {
-	case dstType.AssignableTo(src.Type()):
-		dst.Set(src)
-	case dstType.Kind() == reflect.Array && srcType.Kind() == reflect.Slice:
-		if !dstType.Elem().AssignableTo(r_byte) {
-			return fmt.Errorf("abi: cannot unmarshal %v in to array of elem %v", src.Type(), dstType.Elem())
-		}
-
-		if dst.Len() < output.Type.SliceSize {
-			return fmt.Errorf("abi: cannot unmarshal src (len=%d) in to dst (len=%d)", output.Type.SliceSize, dst.Len())
-		}
-		reflect.Copy(dst, src)
-	case dstType.Kind() == reflect.Interface:
-		dst.Set(src)
-	default:
-		return fmt.Errorf("abi: cannot unmarshal %v in to %v", src.Type(), dst.Type())
-	}
 	return nil
 }
 

--- a/accounts/abi/abi_test.go
+++ b/accounts/abi/abi_test.go
@@ -289,6 +289,37 @@ func TestSimpleMethodUnpack(t *testing.T) {
 	}
 }
 
+func TestUnpackSetInterfaceSlice(t *testing.T) {
+	var (
+		var1 = new(uint8)
+		var2 = new(uint8)
+	)
+	out := []interface{}{var1, var2}
+	abi, err := JSON(strings.NewReader(`[{"type":"function", "name":"ints", "outputs":[{"type":"uint8"}, {"type":"uint8"}]}]`))
+	if err != nil {
+		t.Fatal(err)
+	}
+	marshalledReturn := append(pad([]byte{1}, 32, true), pad([]byte{2}, 32, true)...)
+	err = abi.Unpack(&out, "ints", marshalledReturn)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if *var1 != 1 {
+		t.Errorf("expected var1 to be 1, got", *var1)
+	}
+	if *var2 != 2 {
+		t.Errorf("expected var2 to be 2, got", *var2)
+	}
+
+	out = []interface{}{var1}
+	err = abi.Unpack(&out, "ints", marshalledReturn)
+
+	expErr := "abi: cannot marshal in to slices of unequal size (require: 2, got: 1)"
+	if err == nil || err.Error() != expErr {
+		t.Error("expected err:", expErr, "Got:", err)
+	}
+}
+
 func TestPack(t *testing.T) {
 	for i, test := range []struct {
 		typ string

--- a/accounts/abi/reflect.go
+++ b/accounts/abi/reflect.go
@@ -16,7 +16,10 @@
 
 package abi
 
-import "reflect"
+import (
+	"fmt"
+	"reflect"
+)
 
 // indirect recursively dereferences the value until it either gets the value
 // or finds a big.Int
@@ -61,4 +64,34 @@ func mustArrayToByteSlice(value reflect.Value) reflect.Value {
 	slice := reflect.MakeSlice(reflect.TypeOf([]byte{}), value.Len(), value.Len())
 	reflect.Copy(slice, value)
 	return slice
+}
+
+// set attempts to assign src to dst by either setting, copying or otherwise.
+//
+// set is a bit more lenient when it comes to assignment and doesn't force an as
+// strict ruleset as bare `reflect` does.
+func set(dst, src reflect.Value, output Argument) error {
+	dstType := dst.Type()
+	srcType := src.Type()
+
+	switch {
+	case dstType.AssignableTo(src.Type()):
+		dst.Set(src)
+	case dstType.Kind() == reflect.Array && srcType.Kind() == reflect.Slice:
+		if !dstType.Elem().AssignableTo(r_byte) {
+			return fmt.Errorf("abi: cannot unmarshal %v in to array of elem %v", src.Type(), dstType.Elem())
+		}
+
+		if dst.Len() < output.Type.SliceSize {
+			return fmt.Errorf("abi: cannot unmarshal src (len=%d) in to dst (len=%d)", output.Type.SliceSize, dst.Len())
+		}
+		reflect.Copy(dst, src)
+	case dstType.Kind() == reflect.Interface:
+		dst.Set(src)
+	case dstType.Kind() == reflect.Ptr:
+		return set(dst.Elem(), src, output)
+	default:
+		return fmt.Errorf("abi: cannot unmarshal %v in to %v", src.Type(), dst.Type())
+	}
+	return nil
 }

